### ISSUE TITLE
Add explicit duration- and event-defined window starts to the multitask AR model

### ISSUE THREAD

**Issue by gkondas** [mmcdermott/EveryQuery-conditional#24]
## Summary

Extend the multitask sampling pipeline, dataset, and planned [`ConditionalMultitaskARModel`](https://github.com/mmcdermott/EveryQuery-conditional/issues/23) so each prediction window can have an explicit start rather than always beginning at `prediction_time`.

A multitask window should be defined by two independent specifications:

```text
start:
    prediction_time + start_duration
    OR
    first occurrence of start_event strictly after prediction_time

end:
    resolved_start + duration
    OR
    first occurrence of bound_event strictly after resolved_start
```

For context `i`, window position `k`, and vocabulary code `v`:

```text
target[i, k, v] =
    resolved_start[i, k]
    < occurrence_time(v)
    < resolved_end[i, k]
```

Both endpoints remain open.

This issue applies only to:

- the all-vocabulary multitask sampler from #20;
- the conditioning-code additions from #22;
- the decoder-only `ConditionalMultitaskARModel` from #23.

The existing scalar query sampler, scalar datasets, and `ConditionalQueryARModel` are out of scope.

## Motivation

The current multitask sampler represents each window using only an end specification:

```text
(prediction_time, prediction_time + duration)
(prediction_time, next occurrence of bound_event)
```

The start is implicitly fixed at `prediction_time`.

We also need multitask windows such as:

```text
7 to 37 days after prediction time
the 30 days following the next hospital admission
from the next admission until the next discharge
from 14 days after prediction time until the next ICU admission
```

The [`windowed-queries`](https://github.com/mmcdermott/EQ-motor-gap/tree/windowed-queries/eq_forks) branch in `EQ-motor-gap` implemented part of this behavior for duration-defined starts. In particular, [`windowed_model.py`](https://github.com/mmcdermott/EQ-motor-gap/blob/windowed-queries/eq_forks/multitask/src/every_query/model/windowed_model.py) treats the end duration as relative to the start:

```text
start = prediction_time + S
end   = start + E
```

This issue should incorporate those semantics into the new multitask AR design and extend them to event-defined starts.

## Required window semantics

The start must be resolved before the end.

### Duration-defined start

```text
resolved_start = prediction_time + start_duration
```

A `start_duration` of zero preserves the current behavior:

```text
resolved_start = prediction_time
```

### Event-defined start

```text
resolved_start =
    first occurrence of start_event strictly after prediction_time
```

An occurrence of `start_event` exactly at `prediction_time` does not open the window.

### Duration-defined end

The existing duration becomes relative to the resolved start:

```text
resolved_end = resolved_start + duration
```

For example:

```text
prediction_time = day 0
start_duration  = 7 days
duration        = 30 days

window = (day 7, day 37)
```

The end must not be interpreted as day 30.

### Event-defined end

```text
resolved_end =
    first occurrence of bound_event strictly after resolved_start
```

Do not resolve `bound_event` relative to `prediction_time`.

For example:

```text
prediction_time = day 0
start_event occurs at day 10
bound_event occurs at days 8 and 20

resolved_start = day 10
resolved_end   = day 20
window         = (day 10, day 20)
```

The boundary occurrence at day 8 must be ignored.

### Missing start and end events

Preserve the current missing-end behavior:

- If `bound_event` does not occur strictly after the resolved start, set `resolved_end = +∞`.
- The window therefore extends through the remainder of the recorded timeline.

For a missing event-defined start:

- If `start_event` never occurs strictly after `prediction_time`, the window never opens.
- Treat the window as empty and set every vocabulary target for that window to `False`.
- Do not fall back to beginning the window at `prediction_time`.
- Avoid integer overflow when combining an unresolved start with a duration-defined end.

### Open endpoints

The answer remains:

```text
True iff resolved_start < occurrence_time < resolved_end
```

Therefore:

- the event opening an event-defined window is excluded;
- all events co-timestamped with the start event are excluded;
- events at a duration endpoint are excluded;
- the event closing an event-defined window is excluded;
- all events co-timestamped with the end event are excluded;
- if `start_event == bound_event`, the start is the first future occurrence and the end is the next strictly later occurrence.

## Metadata representation

Extend `MultitaskBoundarySchema` with:

```text
start_durations: List[float32], exactly K entries
start_events:    List[string | null], exactly K entries
```

For each window position, exactly one start representation is active:

```text
start_duration >= 0 and start_event is null
```

or:

```text
start_duration == EVENT_BOUND_DURATION_SENTINEL
and start_event is non-null
```

Keep the existing end fields:

```text
durations:    List[float32], exactly K entries
bound_events: List[string | null], exactly K entries
```

Their semantics become:

```text
duration    = number of days after resolved_start
bound_event = first occurrence strictly after resolved_start
```

For backward compatibility, multitask rows without start columns should be interpreted as:

```text
start_durations = [0.0] * K
start_events    = [null] * K
```

The packed target sidecar remains:

```text
(rows, K, ceil(V / 8)) uint8
bitorder = little
```

## Multitask sampler changes

Update [`sample_multitask_sequences.py`](https://github.com/mmcdermott/EveryQuery-conditional/blob/feat/multitask-boundary-sampler/src/every_query/generate_tasks/sample_multitask_sequences.py) so every one of the `K` windows receives a sampled start specification.

The sampler should cover all combinations (make sure the test suite includes all of these):

| Start             | End      |
| ----------------- | -------- |
| prediction time   | duration |
| prediction time   | event    |
| positive duration | duration |
| positive duration | event    |
| event             | duration |
| event             | event    |

Add configuration along the lines of:

```yaml
# Total probability of an event-defined start.
eventstart_fraction: 0.2

# Total probability that the start remains prediction_time.
# Remaining probability is assigned to a positive duration/ next event start.
prediction_time_start_fraction: 0.4

start_duration_min: 1
start_duration_max: 180
start_duration_distribution: log-uniform

# null => all base observable non-PAD codes
start_event_codes: null
```

Require:

```text
eventstart_fraction >= 0
prediction_time_start_fraction >= 0
eventstart_fraction + prediction_time_start_fraction <= 1
```

Use separate deterministic RNG streams for:

- patient contexts;
- start forms;
- start durations;
- start-event codes;
- end forms;
- end durations;
- end-event codes;
- conditioning codes from #22.

Changing the start distribution must not perturb context, end, or conditioning-code sampling.

The sampler should report:

```text
number of event-defined starts
fraction of event-defined starts that do not resolve
fraction of empty windows
mean positives per window
```

Uniformly sampling start events from the entire vocabulary may produce many missing starts and therefore many all-negative windows. Run a pilot to measure this. If the rate is excessive, sample start-event codes from the context’s observed future events instead. Any such sampling must remain deterministic and must occur in the sampler, not in the dataset or model.

## Start and end resolution

The labeling path should compute:

```text
start_times: (N, K)
end_times:   (N, K)
```

in that order:

```python
start_times = resolve_start_times(
    table,
    subject_ids,
    prediction_times,
    start_durations,
    start_code_indices,
)

end_times = resolve_end_times(
    table,
    subject_ids,
    start_times,
    durations,
    bound_code_indices,
)
```

Both event lookups must use strict next-occurrence semantics equivalent to:

```python
np.searchsorted(times, reference_time, side="right")
```

For duration-defined ends, use checked or saturating arithmetic so that:

```text
INF + duration = INF
```

rather than overflowing.

## Interval-table labeling changes

The current interval-range algorithm assumes that all `K` boundaries for a context share one lower endpoint: `prediction_time`.

That assumption no longer holds. Each `(context, window_position)` may have a different resolved start.

It would be incorrect to continue sorting only by:

```text
subject_id, prediction_time
```

for labeling.

Flatten the `(N, K)` windows into `N × K` logical rows:

```text
subject_id
resolved_start
resolved_end
context_row
window_position
```

Sort these logical windows by:

```text
subject_id, resolved_start, context_row, window_position
```

Run the interval-range labeling kernel using `resolved_start` as the lookup time, then scatter the packed results back into:

```text
(N, K, ceil(V / 8))
```

This should remain one shared flattened-window labeling path. Do not run `K` unrelated full-vocabulary pipelines.

The bounded-memory guarantees from #20 must remain intact.

## Dataset and batch changes

Extend `MultitaskBoundaryBatch` with:

```python
q_start_durations: torch.FloatTensor  # (B, K)
q_start_codes: torch.LongTensor       # (B, K), 0 for duration starts
```

Continue using:

```python
q_durations: torch.FloatTensor        # (B, K), relative end durations
q_bound_codes: torch.LongTensor       # (B, K), 0 for duration ends
```

The dataset should:

1. Load the sampler-materialized start and end specifications.
2. Require exactly `K` entries in every list.
3. Map `start_events` through the same base vocabulary used for targets and boundary events.
4. Reject unknown or PAD start-event codes.
5. Verify that exactly one start representation and one end representation is active per window.
6. Interpret missing legacy start columns as zero-duration starts.
7. Perform no random start sampling in `__getitem__` or `collate`.

The conditioning answers from #22 remain:

```python
condition_answers[i, j] = targets[
    i,
    j,
    vocab_index(condition_codes[i, j]),
]
```

They must therefore be derived from the target for the fully resolved start/end window.

## `ConditionalMultitaskARModel` changes

Generalize the `D_i` token proposed in #23 into a complete window-specification token `W_i`.

The decoder-only sequence becomes:

```text
[patient events,
 W1, C1, A1,
 W2, C2, A2,
 ...
 W(K-1), C(K-1), A(K-1),
 WK]
```

At `W_i`, the model predicts all vocabulary codes occurring inside window `i`.

The causal visibility requirements from #23 remain unchanged:

```text
W_i can see:
    all patient events
    W_i itself
    every W_j, C_j, A_j for j < i

W_i cannot see:
    C_i or A_i
    later windows or conditions
```

No additional start token is required. Construct one window embedding from separate start and end components:

```text
start_spec =
    start_duration_mlp(start_duration)
    OR code_embedding[start_event] + start_marker

end_spec =
    duration_mlp(duration)
    OR code_embedding[bound_event] + bound_marker

window_embedding =
    start_spec
    + end_spec
    + window_token_type_embedding
    + block_position_embedding
```

The start and end projections/markers must be role-specific so the model can distinguish:

```text
start 30 days after prediction time
```

from:

```text
end 30 days after the start
```

Patient codes, conditioning codes, start-event codes, and end-event codes must all use the same live input-embedding table. This preserves compatibility with future ontology-aware embedding replacement.

All `W`, `C`, and `A` tokens still describe questions asked at `prediction_time`. Explicit future starts must be encoded in `W_i`; they must not advance the clinical-time RoPE position. Query-stream tokens should continue to repeat the final patient event’s clinical-time position.

The tied all-vocabulary output head from #23 remains unchanged:

```text
Z_W:   (B, K, H)
E:     (V, H)
logits = Z_W @ E.T + bias
```

## Manifest and restart safety

Increment the multitask format version and record at least:

```text
window_semantics: open_open
start_reference: prediction_time
duration_end_reference: resolved_start
missing_event_start: empty_window
missing_event_end: infinity
event_bound_duration_sentinel
```

The configuration fingerprint must include:

- start-form probabilities;
- start-duration distribution and bounds;
- start-event code pool;
- all existing end sampling parameters;
- the revised window semantics.

Outputs produced with different start semantics must not be reused.

## Required semantic tests

Given:

```text
prediction_time = day 0

events:
    B at day 3
    A at day 5
    X at day 5
    Y at day 6
    B at day 8
    Z at day 8
```

For:

```text
start_event = A
bound_event = B
```

the resolved window is:

```text
(day 5, day 8)
```

The labels must be:

```text
A = False
X = False
Y = True
B = False
Z = False
```

Also test:

- `start_duration = 0` reproduces existing labels;
- positive duration starts resolve relative to `prediction_time`;
- duration ends resolve relative to `resolved_start`;
- event starts use the first occurrence strictly after `prediction_time`;
- event ends use the first occurrence strictly after `resolved_start`;
- an end event before or at the start is ignored;
- co-timestamped start events are excluded;
- co-timestamped end events are excluded;
- a missing start event produces an empty all-false window;
- a missing end event produces an unbounded window;
- equal start and end event codes select consecutive occurrences;
- all six start/end combinations are supported;
- `INF + duration` does not overflow.

## Kernel and integration tests

- Randomized comparison with a naive `(start, end)` oracle.
- Different start times across the `K` positions of one context.
- Resolved-start ordering differs from prediction-time ordering.
- Flatten/sort/scatter output matches the naive oracle.
- Chunked and unchunked results match.
- Packed row/window alignment is preserved.
- Existing zero-start fixtures remain identical after unpacking.
- Start fields survive metadata loading and global shuffle.
- Dataset collation returns the expected `(B, K)` start tensors.
- Conditioning answers are extracted from the correct resolved windows.
- Window embeddings select the correct start and end branches.
- Event-start and event-end markers are distinguishable.
- Multitask causal visibility remains unchanged.
- Clinical-time RoPE positions remain at prediction time.

## Expected repository changes

Start by branching from the multitask-boundary-sampler branch

Update:

```text
src/every_query/generate_tasks/sample_multitask_sequences.py
src/every_query/generate_tasks/interval_table.py
src/every_query/generate_tasks/configs/sample_multitask_sequences_config.yaml
src/every_query/data/schema.py
src/every_query/data/multitask_dataset.py
```

Update the `ConditionalMultitaskARModel` introduced by #23 and its tests.

Do not modify the scalar query sampler, `QuerySeqSchema`, scalar dataset, or `ConditionalQueryARModel` as part of this issue.

## Acceptance criteria

- [ ] Every multitask window has an explicit start and end specification.
- [ ] Starts can be prediction time, positive durations, or future events.
- [ ] Duration ends are relative to the resolved start.
- [ ] Event ends are resolved strictly after the resolved start.
- [ ] Both endpoints remain open.
- [ ] Missing event starts produce empty all-false windows.
- [ ] Missing event ends retain the existing `+∞` behavior.
- [ ] Equal start/end event codes use consecutive occurrences.
- [ ] All six start/end combinations are sampled and labeled correctly.
- [ ] Different starts across the `K` windows are handled correctly by the optimized labeler.
- [ ] Packed-output and bounded-memory guarantees from #20 remain intact.
- [ ] The dataset only loads and encodes sampler-materialized specifications.
- [ ] Conditioning answers from #22 use the correct resolved windows.
- [ ] `ConditionalMultitaskARModel` represents each complete window with `W_i`.
- [ ] The causal sequence and tied vocabulary head from #23 remain unchanged.
- [ ] Explicit future starts do not alter clinical-time RoPE positions.
- [ ] Legacy multitask rows without start fields retain prediction-time-start semantics.
- [ ] The scalar AR pipeline is unchanged.
- [ ] Manifest fingerprinting prevents incompatible output reuse.
- [ ] Unit, differential, dataset, and model tests pass.
